memcached.c

```diff
@@ -4136,18 +4209,62 @@ static int _meta_flag_preparse(token_t *tokens, const size_t ntokens, struct _me
                 break;
             // mset-related.
             case 'F':
+                if (!safe_strtoul(tokens[i].value+1, &of->client_flags)) {
+                    of->has_error = true;
+                }
+                break;
             case 'S':
-            case 'C':
+                if (!safe_strtol(tokens[i].value+1, &tmp_int)) {
+                    of->has_error = true;
+                } else {
+                    // Size is adjusted for underflow or overflow once the
+                    // \r\n terminator is added.
+                    if (tmp_int < 0 || tmp_int > (INT_MAX - 2)) {
+                        *errstr = "CLIENT_ERROR invalid length";
+                        of->has_error = true;
+                    } else {
+                        of->value_len = tmp_int + 2; // \r\n
+                    }
+                }
+                break;
+            case 'C': // mset, mdelete, marithmetic
+                if (!safe_strtoull(tokens[i].value+1, &of->req_cas_id)) {
+                    *errstr = "CLIENT_ERROR bad token in command line format";
+                    of->has_error = true;
+                } else {
+                    of->has_cas = true;
+                }
+                break;
+            case 'M': // mset and marithmetic mode switch
+                if (tokens[i].length != 2) {
+                    *errstr = "CLIENT_ERROR incorrect length for M token";
+                    of->has_error = 1;
+                } else {
+                    of->mode = tokens[i].value[1];
+                }
+                break;
+            case 'J': // marithmetic initial value
+                if (!safe_strtoull(tokens[i].value+1, &of->initial)) {
+                    *errstr = "CLIENT_ERROR invalid numeric initial value";
+                    of->has_error = 1;
+                }
+                break;
+            case 'D': // marithmetic delta value
+                if (!safe_strtoull(tokens[i].value+1, &of->delta)) {
```

Comment:
So the `M` mode option for `ma` accepts `I, `D`, `+` or `-`,  and assumes implicitly that the `D` option will not accept negative values (which `safe_strtoull` handles and appropriately fails).

But the protocol description does say that the argument for `-D` is an `unsigned int`, but `safestrtoull` will happily accept a string value > an `unsigned int`. Maybe this is not that important, but we could easily adapt the protocol documentation to just say `will accept an argument that is a positive unsigned long long` maybe? 🤷‍♂

Reply:
Nice catch! This was supposed to match the incr/decr documentation as it handles arguments the same way. I've updated a few words to explicitly say 64-bit unsigned integer.